File: lint-rules-android-lint/src/test/java/com/vanniktech/lintrules/android/InvalidSingleLineCommentDetectorTest.java
package com.vanniktech.lintrules.android;

import com.android.tools.lint.checks.infrastructure.LintDetectorTest;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Issue;
import java.util.List;
import org.intellij.lang.annotations.Language;

import static com.vanniktech.lintrules.android.AndroidDetectorTest.NO_WARNINGS;
import static java.util.Collections.singletonList;
import static org.fest.assertions.api.Assertions.assertThat;

public class InvalidSingleLineCommentDetectorTest extends LintDetectorTest {
  public void testInvalidSingleLineCommentNoSpace() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    //Something.\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo("src/foo/Example.java:5: Warning: Comment does not contain a space at the beginning. [InvalidSingleLineComment]\n"
        + "    //Something.\n"
        + "      ^\n"
        + "0 errors, 1 warnings\n");
  }

  public void testInvalidSingleLineCommentNotStartingCapitalLetter() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    // something.\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo("src/foo/Example.java:5: Warning: Comments first word should be capitalized. [InvalidSingleLineComment]\n"
        + "    // something.\n"
        + "       ^\n"
        + "0 errors, 1 warnings\n");
  }

  public void testInvalidSingleLineCommentNoPeriodAtEnd() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    // Something\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo("src/foo/Example.java:5: Warning: Comment does not end with a period. [InvalidSingleLineComment]\n"
        + "    // Something\n"
        + "    ~~~~~~~~~~~~\n"
        + "0 errors, 1 warnings\n");
  }

  public void testInvalidSingleLineCommentEndingWithPeriod() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    // Something. Do not modify!\n"
        + "    // Something. Do not modify.\n"
        + "    // Something. Do not modify?\n"
        + "    // Something. (Do not modify)\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testInvalidSingleLineCommentNoSpaceBefore() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    int foo = 5 + 5;// Something.\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo("src/foo/Example.java:5: Warning: Comment does not start with a single space.. [InvalidSingleLineComment]\n"
        + "    int foo = 5 + 5;// Something.\n"
        + "                   ~\n"
        + "0 errors, 1 warnings\n");
  }

  public void testInvalidSingleLineCommentIgnoresLinks() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    String link1 = \"https://android.com/\";\n"
        + "    String link2 = \"http://android.com/\";\n"
        + "    String link3 = \"market://details?id=5\";\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testInvalidSingleLineCommentIgnoresNoPmd() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    // NOPMD\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testInvalidSingleLineCommentIgnoresNoInspection() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    //noinspection\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testInvalidSingleLineCommentLinkInComment() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    // http://stackoverflow.com/a/38480079\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testInvalidSingleLineCommentIgnoresJustComment() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    //\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testInvalidSingleLineCommentIgnoresJustCommentWithTrailingWhitespace() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    // \n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo("src/foo/Example.java:5: Warning: Comment contains trailing whitespace. [InvalidSingleLineComment]\n"
        + "    // \n"
        + "    ~~~\n"
        + "0 errors, 1 warnings\n");
  }

  public void testInvalidSingleLineCommentTrailingWhitespace() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import android.content.res.Resources;\n"
        + "public class Example {\n"
        + "  public void foo() {\n"
        + "    // Something. \n"
        + "  }\n"
        + "}";
    assertThat(lintProject(java(source))).isEqualTo("src/foo/Example.java:5: Warning: Comment contains trailing whitespace. [InvalidSingleLineComment]\n"
        + "    // Something. \n"
        + "    ~~~~~~~~~~~~~~\n"
        + "0 errors, 1 warnings\n");
  }

  @Override protected Detector getDetector() {
    return new InvalidSingleLineCommentDetector();
  }

  @Override protected List<Issue> getIssues() {
    return singletonList(InvalidSingleLineCommentDetector.ISSUE_INVALID_SINGLE_LINE_COMMENT);
  }

  @Override protected boolean allowCompilationErrors() {
    return false;
  }
}

File: lint-rules-android-lint/src/test/java/com/vanniktech/lintrules/android/InvalidSingleLineCommentDetectorTest.kt
package com.vanniktech.lintrules.android

import com.android.tools.lint.checks.infrastructure.TestFiles.gradle
import com.android.tools.lint.checks.infrastructure.TestFiles.java
import com.android.tools.lint.checks.infrastructure.TestLintTask.lint
import com.vanniktech.lintrules.android.InvalidSingleLineCommentDetector.ISSUE_INVALID_SINGLE_LINE_COMMENT
import org.junit.Test

class InvalidSingleLineCommentDetectorTest {
  @Test fun invalidSingleLineCommentNoSpace() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    //Something.
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expect("""
          |src/foo/Example.java:5: Warning: Comment does not contain a space at the beginning. [InvalidSingleLineComment]
          |    //Something.
          |    ~~~
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for src/foo/Example.java line 4: Add space:
          |@@ -5 +5
          |-     //Something.
          |+     // Something.
          |""".trimMargin())
  }

  @Test fun invalidSingleLineCommentNotStartingCapitalLetter() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    // something.
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expect("""
          |src/foo/Example.java:5: Warning: Comments first word should be capitalized. [InvalidSingleLineComment]
          |    // something.
          |       ^
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for src/foo/Example.java line 4: Capitalized first word:
          |@@ -5 +5
          |-     // something.
          |+     // Something.
          |""".trimMargin())
  }

  @Test fun invalidSingleLineCommentNoPeriodAtEnd() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    // Something
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expect("""
          |src/foo/Example.java:5: Warning: Comment does not end with a period. [InvalidSingleLineComment]
          |    // Something
          |    ~~~~~~~~~~~~
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for src/foo/Example.java line 4: Add period:
          |@@ -5 +5
          |-     // Something
          |+     // Something.
          |""".trimMargin())
  }

  @Test fun validSingleLineCommentEndingWithPeriod() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    // Something. Do not modify!
          |    // Something. Do not modify.
          |    // Something. Do not modify?
          |    // Something. (Do not modify)
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expectClean()
  }

  @Test fun invalidSingleLineCommentNoSpaceBefore() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    int foo = 5 + 5;// Something.
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expect("""
          |src/foo/Example.java:5: Warning: Comment declaration is not preceded by a single space. [InvalidSingleLineComment]
          |    int foo = 5 + 5;// Something.
          |                   ~~~
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for src/foo/Example.java line 4: Add space:
          |@@ -5 +5
          |-     int foo = 5 + 5;// Something.
          |+     int foo = 5 + 5; // Something.
          |""".trimMargin())
  }

  @Test fun invalidSingleLineCommentIgnoresLinks() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    String link1 = "https://android.com/";
          |    String link2 = "http://android.com/";
          |    String link3 = "market://details?id=5";
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expectClean()
  }

  @Test fun invalidSingleLineCommentIgnoresNoPmd() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    // NOPMD
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expectClean()
  }

  @Test fun invalidSingleLineCommentAllowsStartingDigit() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    // 2 plus 4 equals 6.
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expectClean()
  }

  @Test fun invalidSingleLineCommentIgnoresNoInspection() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    //noinspection
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expectClean()
  }

  @Test fun invalidSingleLineCommentLinkInComment() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    // http://stackoverflow.com/a/38480079
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expectClean()
  }

  @Test fun invalidSingleLineCommentIgnoresJustComment() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    //
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expectClean()
  }

  @Test fun invalidSingleLineCommentIgnoresJustCommentWithTrailingWhitespace() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    //
          |  }
          |}""".trimMargin().replace("//", "// ")))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expect("""
          |src/foo/Example.java:5: Warning: Comment contains trailing whitespace. [InvalidSingleLineComment]
          |    //
          |    ~~~
          |0 errors, 1 warnings""".trimMargin().replace("//", "// "))
        .expectFixDiffs("""
          |Fix for src/foo/Example.java line 4: Remove trailing whitespace:
          |@@ -5 +5
          |-     //
          |+     //
          |""".trimMargin().replace("-     //", "-     // "))
  }

  @Test fun invalidSingleLineCommentTrailingWhitespace() {
    lint()
        .files(java("""
          |package foo;
          |import android.content.res.Resources;
          |class Example {
          |  public void foo() {
          |    // Something.
          |  }
          |}""".trimMargin().replace("// Something.", "// Something. ")))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expect("""
          |src/foo/Example.java:5: Warning: Comment contains trailing whitespace. [InvalidSingleLineComment]
          |    // Something.
          |    ~~~~~~~~~~~~~~
          |0 errors, 1 warnings""".trimMargin().replace("// Something.", "// Something. "))
        .expectFixDiffs("""
          |Fix for src/foo/Example.java line 4: Remove trailing whitespace:
          |@@ -5 +5
          |-     // Something.
          |+     // Something.
          |""".trimMargin().replace("-     // Something.", "-     // Something. "))
  }

  @Test fun worksOnGradleFiles() {
    lint()
        .files(gradle("""
          |buildscript {
          |  repositories {
          |    mavenCentral() // we need this.
          |  }
          |}""".trimMargin()))
        .issues(ISSUE_INVALID_SINGLE_LINE_COMMENT)
        .run()
        .expect("""
          |build.gradle:3: Warning: Comments first word should be capitalized. [InvalidSingleLineComment]
          |    mavenCentral() // we need this.
          |                      ^
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for build.gradle line 2: Capitalized first word:
          |@@ -3 +3
          |-     mavenCentral() // we need this.
          |+     mavenCentral() // We need this.
          |""".trimMargin())
  }
}
